### ISSUE TITLE
"packed" attribute missing on nested structs 

### ISSUE THREAD

**Issue by mntmn** [Chadderz121/csud#6]
I struggled many hours today to find out why the csud driver would initialize correctly but then fail to iterate the root hub. The problem was that since GCC4.4(?) the packed attribute has to be set on nested structs too if they need to be bit-packed. Example:

```
struct UsbConfigurationDescriptor {
    u8 DescriptorLength; // +0x0
    enum DescriptorType DescriptorType : 8; // +0x1
    u16 TotalLength; // +0x2
    u8 InterfaceCount; // +0x4
    u8 ConfigurationValue; // +0x5
    u8 StringIndex; // +0x6
    struct {
        unsigned _reserved0_4 : 5; // @0
        bool RemoteWakeup : 1; // @5
        bool SelfPowered : 1; // @6
        unsigned _reserved7 : 1; // @7
    } Attributes; // +0x7
    u8 MaximumPower; // +0x8
} __attribute__ ((__packed__));
```

sizeof(struct UsbConfigurationDescriptor) returns 12 for me, not 9 as the code implies. This breaks the descriptor discovery loop in UsbConfigure() because of wrong offsets.

Adding the packed attribute on the inner struct "Attributes" fixed the problem for me:

```
...
    struct {
        unsigned _reserved0_4 : 5; // @0
        bool RemoteWakeup : 1; // @5
        bool SelfPowered : 1; // @6
        unsigned _reserved7 : 1; // @7
    } __attribute__ ((__packed__)) Attributes; // +0x7
    u8 MaximumPower; // +0x8
} __attribute__ ((__packed__));
```

For reference, I am running GCC 4.7.4:
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 4.7.4 20130913 (release) [ARM/embedded-4_7-branch revision 202601]

Cheers,
Lukas

**Comment by Chadderz121** [comment]
Good find, sorry for that mistake! I'm actually rather suprised the use of bit fields doesn't imply packed somehow, it seems quite illogical to not have packed bitfields. I'm sure there must be some good reason for not doing, but either way it's an easy fix!

**Comment by Chadderz121** [comment]
Fixed as of commit bfbdbc23389a24d81fe2c80c2aa6b5480ea8d1f8.
